### ISSUE TITLE
Add option to observe despite missing addons

### ISSUE THREAD

**Issue by ProditorMagnus** [wesnoth/wesnoth#4358]
Re #2747
Re https://github.com/ProditorMagnus/Ageless-for-1-14/issues/7
Split from #2324

If user attempts to observe with invalid configuration such has having higher addon version than host, then the error popup should have button to ignore warning and try to observe anyways.

**Comment by ProditorMagnus** [comment]
Relevant code https://github.com/wesnoth/wesnoth/blob/60fd504caf46bbb1648de369303b76225d805e5d/src/game_initialization/lobby_data.cpp#L513-L532

**Comment by gfgtdf** [comment]
I don't think this is really a good idea, observing a game without the addon instlled will usually casue OOS, of if there is such a button it's quite likely that at least one person that doesnt know what he is doing will click it. Even if we put a fat warning on it.

**Comment by ProditorMagnus** [comment]
Having OOS as observer is mostly harmless though.

**Comment by gfgtdf** [comment]
> Having OOS as observer is mostly harmless though.

Hmm yes but oversver can still be assigned sides, if for exampel when a player leaves.

**Comment by ProditorMagnus** [comment]
For my use case it would be enough to limit scope to only cases where observer has addon, but it is higher version. This situation is mostly limited to some days after update, and joining and telling about it in game is useful in getting host to update too.

**Comment by soliton-** [comment]
> if there is such a button it's quite likely that at least one person that doesnt know what he is doing will click it. Even if we put a fat warning on it.

I'd rather inconvenience that person that clicks through everything without reading than people that know what they're doing (for example addon authors like ProditorMagnus).

**Comment by jostephd** [comment]
@ProditorMagnus First of all, the next time you open an issue, please don't just say what the proposed change is and where the code that needs to change is, but also say what the use-case is, as you did here:

> For my use case it would be enough to limit scope to only cases where observer has addon, but it is higher version. This situation is mostly limited to some days after update, and joining and telling about it in game is useful in getting host to update too.

Can we consider other solutions to this? For example, have wesnothd talk to campaignd and inform a player who hosts a game with old versions of add-ons? Or, say, support having multiple versions of an add-on installed, and choosing which one to read into the configuration based on what version the host uses?

And even if we _did_ want to allow observing with different versions, the decision shouldn't be the user's. The decision should be made by the addon author. We could even require all addons to use [semver](https://www.semver.org/) version numbering and infer compatibility from that, without needing a new pbl field.

> I'd rather inconvenience that person that clicks through everything without reading than people that know what they're doing (for example addon authors like ProditorMagnus).

The fact that @ProditorMagnus wants to tell people to upgrade is not a good reason to show players a dialog which essentially asks them if they'd like to get an OOS error. There is any number of other ways to achieve @ProditorMagnus's goal without confusing users and giving them the impression that wesnoth is an unstable game, and it is not good UI design to show the user a dialog and expect them to read through it when showing the dialog can be avoided in the first place.

**Comment by gfgtdf** [comment]
> And even if we did want to allow observing with different versions, the decision shouldn't be the user's. The decision should be made by the addon author

In fact @ProditorMagnus this is already possible, you can just use use min_compatible_addon_version (not sure whether that's the exact name) if you don't want it to apply to others you can quard if in some #ifhave

**Comment by gfgtdf** [comment]
> Can we consider other solutions to this? For example, have wesnothd talk to campaignd and inform a player who hosts a game with old versions of add-ons? 

Independently of ravanas usecase this is a good idea

**Comment by ProditorMagnus** [comment]
I agree even. I do not really need the functionality, for Linux I have made the change there already, I was just complaining about it when I was using windows and then was told to create issue for this.

> have wesnothd talk to campaignd and inform a player who hosts a game with old versions of add-ons  

Would be nice
> support having multiple versions of an add-on installed  

Seems like too complex development for the benefit
> the decision should be made by the addon author  

Having separate addon_min_version about observing might work
> require all addons to use semver version numbering  

Sounds like unfeasible to enforce

**Comment by ProditorMagnus** [comment]
> In fact @ProditorMagnus this is already possible, you can just use use min_compatible_addon_version (not sure whether that's the exact name) if you don't want it to apply to others you can quard if in some #ifhave

That is quite good idea.

**Comment by soliton-** [comment]
> The fact that @ProditorMagnus wants to tell people to upgrade is not a good reason to show players a dialog which essentially asks them if they'd like to get an OOS error. There is any number of other ways to achieve @ProditorMagnus's goal without confusing users and giving them the impression that wesnoth is an unstable game, and it is not good UI design to show the user a dialog and expect them to read through it when showing the dialog can be avoided in the first place.

So how do you avoid showing the dialog? If the dialog alone gives the impression of an unstable game then we need to remove the addon feature to fix this.

To me the issue is not why exactly someone wants to join the game anyway. The issue is that there is no technical reason that they can't. So there is no way for software to decide whether they shouldn't. Even more so because we know the current check has bugs.

If it's not enough to default to not join the game and we can't expect our users to be able to read and understand the question about joining anyway then we can hide the question behind debug mode or so. Requiring people to modify the source code is not acceptable IMO.

Btw, we already ask players if they want more OOS errors in the OOS dialog itself. I also find it a stretch to imagine someone being confused by the simple statement that they're missing an addon.

**Comment by jostephd** [comment]
> have wesnothd talk to campaignd and inform a player who hosts a game with old versions of add-ons?

#4388

> Sounds like unfeasible to enforce

"Enforce"? We already have a `version_info` class. It won't be hard to compare two version numbers in the right way. We can't police add-on authors for setting the version number correctly, but we don't need to, either. If it's a concern, we could have a new pbl field to opt-in to treating the version number as indicating compatibility in the semver sense. I guess we'd need to provide some way for addon authors to know what's the oldest minor version of any other player, though? :thinking: 

> So how do you avoid showing the dialog? If the dialog alone gives the impression of an unstable game then we need to remove the addon feature to fix this.

Nobody is proposing to remove the addons feature.

My point was that if we show the player a message "Hey, joining this game probably not work, do you want to try it anyway?" and they say yes and then they get an OOS error, they're going to tell everyone that wesnoth is unstable. I don't think we should offer players a choice between two options when we _know_ one of the two is going to lead to OOS errors. It might work to hide such a question behind debug mode (that is, `wesnoth --debug`) or an advanced preference, as you suggested - but I don't think we should ask that question by default, especially since Ravana's use-case has other solutions, such as #4388.

> To me the issue is not why exactly someone wants to join the game anyway. The issue is that there is no technical reason that they can't. So there is no way for software to decide whether they shouldn't. Even more so because we know the current check has bugs.

What? I think we might be starting from different assumptions here. My understanding that we're talking about a player who wants to observe a game when wesnoth _knows_ that the player has a different version of some addon (and `addon_min_version` isn't set, so we default to assuming the two versions are incompatible). Are you talking about this situation or another one? And what are the known bugs in the current check, are you thinking of github issues other than this one?

> Btw, we already ask players if they want more OOS errors in the OOS dialog itself.

Indeed, but we don't have any other choice there, do we? Whereas in this case we do have other options.

**Comment by soliton-** [comment]
> I don't think we should offer players a choice between two options when we know one of the two is going to lead to OOS errors.

How do we know that? Perhaps more importantly how do we know the OOS errors are preventing whatever the player is trying to do? (chatting, moderating, debugging, etc)

> My understanding that we're talking about a player who wants to observe a game when wesnoth knows that the player has a different version of some addon (and addon_min_version isn't set, so we default to assuming the two versions are incompatible).

That or any other situation where wesnoth _guesses_ observing might not work well.

> And what are the known bugs in the current check, are you thinking of github issues other than this one?

https://github.com/wesnoth/wesnoth/issues/4055 Don't remember from the top of my head if there are more.

> > Btw, we already ask players if they want more OOS errors in the OOS dialog itself.

> Indeed, but we don't have any other choice there, do we? Whereas in this case we do have other options.

We could prevent the user from continuing to play since we can guess more OOS errors are going to occur.

**Comment by jostephd** [comment]
> How do we know that? Perhaps more importantly how do we know the OOS errors are preventing whatever the player is trying to do? (chatting, moderating, debugging, etc)

By looking at the addon's `addon_min_version`. Suppose the player has addon version X installed and a game is played with addon version Y. If the pbl file of the newer version (whichever of X and Y is newer) doesn't have an `addon_min_version` field that declares interoperability with the older version, we should assume the two versions are _not_ interoperable.

As to chatting/moderating, I don't know, maybe there should be a way to join permanently blindfolded? So you just get the chat, but not the game actions? So the dialog would have three options:

> You have add-on foo version 1 installed. The host has addon foo version 2 installed. The two versions are not compatible (add-on authors: see `addon_min_version` in https://wiki.wesnoth.org/ModificationWML). What do you want to do?
> 
> - Abort joining
> - Join permanently blindfolded (chat only)
> - Try to join normally (OOS errors may be raised during the game)

> That or any other situation where wesnoth _guesses_ observing might not work well.

Can you give an example of an "any other situation"? It's possible different situations will require different default behaviors, for example.

> #4055 Don't remember from the top of my head if there are more.

Thanks. Okay, I see your point: the user should have the final say. I do agree. So there _should_ be a dialog. But it still leaves open the question of whether the dialog's default should be, whether it should be guarded by a preference, etc.

> We could prevent the user from continuing to play since we can guess more OOS errors are going to occur.

Well, yes, but I thought you just finished convincing _me_ that making it impossible to join is a bad idea and the user should have the final say... surely the same applies to whether to continue after an in-game OOS error :) (I'm also thinking of OOS errors during replays)

As an example, I have some old savefiles that have one or two OOS errors in them. I know about those errors, and when I replay those savefiles I just skip past them. But I'd like to be able to do so.

**Comment by soliton-** [comment]
Yes, the OOS example was to show there are other dialogs where we allow the user final say about what to do even if there is a good chance they‘ll encounter errors. I did not mean to suggest to change that.

Those 3 options sound good. Not sure how difficult the chat only option would be to implement. Perhaps blindfold + ignoring OOS errors might be enough. If it’s not easy the other two options would be fine for a start as well.

> Can you give an example of an "any other situation"? It's possible different situations will require different default behaviors, for example.

I rather meant different reasons we might not allow joining for. I guess era/mod missing is all there is? And that bug report dialog should also not prevent joining.

Default being what we assume is the best idea is good. So not joining should be the default in any case, I think.

**Comment by ProditorMagnus** [comment]
Only useful part of the second option would be being protected from being assigned sides.

**Comment by soliton-** [comment]
Also not erroring out of the game because of unknown unit types for example. For that to work blindfold + OOS ignore is probably not though. Ideally chat only mode would ignore everything but chat commands, I‘d say. That probably belongs in its own issue though...

**Comment by ProditorMagnus** [comment]
Only starting unknown units cause kick, those that appear during game are only Lua errors.

**Comment by ProditorMagnus** [comment]
Current use case as workaround for misconfigured addon https://github.com/knyghtmare/War_of_Legends/issues/110

**Comment by Pentarctagon** [comment]
That page gives me a 404.

**Comment by ProditorMagnus** [comment]
I created issue under era since I did not find repository for modifications. I suppose he moved it to correct place, but that is private repository. My profile shows `1 contribution in private repositories` and that could be the only reason from today.

Addon does

```
[modification]
    id=WoL_mod_disable_elvish_sylph
    name= _ "Disable Elvish Sylph"
    description = _ "Disallows Elvish Enchantresses from advancing to Elvish Sylphs."
    type=hybrid
    require_modification=yes
    addon_min_version=1.0.1

    [modify_unit_type]
        type="Elvish Enchantress"
        remove_advancement="Elvish Sylph"
        set_experience=150
    [/modify_unit_type]
[/modification]
```
while [modify_unit_type] does not need require_modification=yes. So to solve it I had to create _info.cfg with
```
#
# File automatically generated by Wesnoth to keep track
# of version information on installed add-ons. DO NOT EDIT!
#
[info]
	core=""
	dependencies=""
	title="WoL_mods"
	type="mod_mp"
	uploads=1
	version="1.0.1"
[/info]
```
and _main.cfg with
```
[modification]
    id=WoL_mod_disable_elvish_sylph
[/modification]
```

**Comment by knyghtmare** [comment]
> I created issue under era since I did not find repository for modifications. I suppose he moved it to correct place, but that is private repository. My profile shows 1 contribution in private repositories and that could be the only reason from today.

That issue you reported was patched and the addon updated.
